### PR TITLE
Add systemctl for systemd based systems

### PR DESCRIPTION
**Pull request by th-2021**
First take on Independence from service #2612

### CODE REVIEW

File: cobbler/utils.py
Comment (SchoolGuy) [inline cobbler/utils.py:240]: `supervisorctl restart <UNIT>`

File: cobbler/utils.py
Comment (SchoolGuy) [inline cobbler/utils.py:242]: This will fail if service is not available. So the `else` branch would be a `logger.warning` that the service was not restarted and then the method returns imho.

**Review by SchoolGuy — COMMENTED** [review]
Please change the branch away from `master`. In case we would merge anything on `origin/master` you would run into trouble. Also the git graph looks a bit better with a feature branch. It is not a must but I would love to see it if it is not too much trouble for you.

File: cobbler/utils.py
Comment (th-2021) [reply cobbler/utils.py]: and for status in check.py? Is is supervisorctl status <service>?

**Comment by th-2021** [comment]
Can we leave it this time with master? Next time I will use a branch.

File: cobbler/actions/check.py
Comment (heroin-moose) [inline cobbler/actions/check.py:138]: The order seems to be wrong. Containers that have supervisorctl still have systemctl. Also is there a reason for separate functions? E.g. a list of binaries with iterator would be much smaller.

Comment (heroin-moose) [reply cobbler/actions/check.py:138]: That's what we have in system-tests:

```
srvctl() {
        $(command -v supervisorctl systemctl | head -1) ${@}
}
```

File: cobbler/utils.py
Comment (SchoolGuy) [reply cobbler/utils.py]: Yes: http://supervisord.org/running.html#supervisorctl-actions

**Comment by SchoolGuy** [comment]
> Can we leave it this time with master? Next time I will use a branch.

@th-2021 Yes it is fine this time. I will freeze the branch until you are done.

File: cobbler/actions/check.py
Comment (th-2021) [reply cobbler/actions/check.py:138]: I corrected the order. I introduced the functions because I need the same for service restart. So instead of having the commands in 2 places I thought it's better to have a function for it (and is_systemd was already there)

File: cobbler/utils.py
Comment (SchoolGuy) [inline cobbler/utils.py:232]: I believe this can be simplified.

```suggestion
    return os.path.exists("/usr/bin/supervisord"):
```

Comment (SchoolGuy) [reply cobbler/utils.py:232]: Another (in my eyes preferred) option would be to use the RPC interface because we don't need to rely on exit codes or stdout parsing.

Docs: http://supervisord.org/api.html

File: cobbler/actions/check.py
Comment (SchoolGuy) [inline cobbler/actions/check.py:139]: I would love if we could use the XMLRPC API for this too.

**Review by SchoolGuy — COMMENTED** [review]
If what I am asking is too much it is fine but that would be my wish-list. Other then that I would like to see the Git History cleaned up before we merge!

File: cobbler/actions/check.py
Comment (th-2021) [reply cobbler/actions/check.py:139]: what do you mean by this? I'm not familiar with xmlrpc.

Comment (SchoolGuy) [reply cobbler/actions/check.py:139]: I am referring to https://github.com/cobbler/cobbler/pull/2841#discussion_r754260340

File: cobbler/utils.py
Comment (SchoolGuy) [inline cobbler/utils.py:239]: Same thing here

```suggestion
    return os.path.exists("/usr/sbin/service")
```

File: cobbler/actions/check.py
Comment (SchoolGuy) [reply cobbler/actions/check.py:139]: I'll start a new discussion for this.

Comment (SchoolGuy) [inline cobbler/actions/check.py:139]: Another (in my eyes preferred) option would be to use the XML-RPC interface because we don't need to rely on exit codes or stdout parsing.

Docs: http://supervisord.org/api.html

This means we would have more Python Code but the detection would be much more reliable imho.

**Review by SchoolGuy — COMMENTED** [review]
So code is overall fine in my eyes expect for the thread I started.

File: cobbler/actions/check.py
Comment (SchoolGuy) [inline cobbler/actions/check.py]: Suggestion from my side to prevent open connections on any problem.

```suggestion
            with ServerProxy('http://localhost:9001/RPC2') as server:
                process_info = server.supervisor.getProcessInfo(which)
                if process_info['statename'] != "RUNNING":
                    status.append("service %s is not running%s" % (which, notes))
                    return
```

File: cobbler/utils.py
Comment (SchoolGuy) [inline cobbler/utils.py]: Same as the other suggestion:

```suggestion
    with ServerProxy('http://localhost:9001/RPC2') as server:
        try:
            server.supervisor.getState()
        except OSError as e:
            return False
        return True
```

File: cobbler/utils.py
Comment (SchoolGuy) [inline cobbler/utils.py:252]: Same as above!

```suggestion
        with ServerProxy('http://localhost:9001/RPC2') as server:
            server.supervisor.stopProcess(service_name)
            server.supervisor.startProcess(service_name)
```

**Review by SchoolGuy — CHANGES_REQUESTED** [review]
This is good stuff! I like that a lot more already! However I believe there is one thing more we need to address.

File: cobbler/utils.py
Comment (SchoolGuy) [inline cobbler/utils.py:233]: Okay Codacy complains correctly that this variable is not used. So let's fix that:
```suggestion
      except OSError:
```

**Comment by SchoolGuy** [comment]
@th-2021 This is really good work! If I may then I would restructure the Git Commits so they look a bit better?

**Review by SchoolGuy — APPROVED** [review]
LGTM in my eyes. I would love to have this here as it would solve a feature request!

**Comment by th-2021** [comment]
do you want a single commit? Or are the 6 ok?

**Comment by SchoolGuy** [comment]
@th-2021 I think a single one with an extended commit message which explains what you did here is enough!

**Comment by SchoolGuy** [comment]
@heroin-moose What is your opinion on this now? Mergable or not? You had reservations to this before.

**Comment by SchoolGuy** [comment]
@th-2021 Please have a look at the commits I added. The first one I would like you to sqash and the second one too if you find it acceptable. I can't do it myself as I can only add commits and not force push to your `master` branch.

**Comment by heroin-moose** [comment]
> @heroin-moose What is your opinion on this now? Mergable or not? You had reservations to this before.

I think It would be better to do the checks on the Cobbler start, there is no need to check again every time. Also renaming rc to ret_code seems out of scope (besides rc being better IMO).

File: cobbler/utils.py
Comment (heroin-moose) [inline cobbler/utils.py:227]: Now it checks for the working supervisord.

File: cobbler/utils.py
Comment (heroin-moose) [inline cobbler/utils.py:264]: I think it's better to clarify that we cannot restart the service because there is no known service manager.

**Comment by SchoolGuy** [comment]
@heroin-moose Where would you hold the data which process manager we use? I think `settings.py` is the wrong place but I don't want to clutter `api.py`/`remote.py` with more things.

**Comment by heroin-moose** [comment]
> @heroin-moose Where would you hold the data which process manager we use? I think `settings.py` is the wrong place but I don't want to clutter `api.py`/`remote.py` with more things.

Hmm... good point.

**Comment by SchoolGuy** [comment]
@heroin-moose IMHO while I agree that we need to save us the CPU cycles of checking on every restart of a service, I believe that this is an improvement we can do later when we know where we want to hold the state. I would say we merge this after your suggestions have been applied and open a new issue where we document our new reservations for the code. This is already an improvement in my eyes big enough to put it in the next release.

**Comment by heroin-moose** [comment]
@SchoolGuy, agree.

**Comment by SchoolGuy** [comment]
@th-2021 I tried addressing @heroin-moose 's wishes. It is your PR so please decide if this is fitting or if you want to change it. If yes please squash again. After that I will merge, so we have this in! :)

**Comment by th-2021** [comment]
Done

**Comment by SchoolGuy** [comment]
@heroin-moose Anything left from your side?

**Comment by heroin-moose** [comment]
Nah, I'm good.

File: cobbler/utils.py
Comment (agraul) [inline cobbler/utils.py:225]:
```suggestion
    Return whether or not this system uses supervisod.
```

File: cobbler/utils.py
Comment (agraul) [inline cobbler/utils.py:239]:
```suggestion
    Return whether or not this system uses service.
```

File: cobbler/utils.py
Comment (agraul) [inline cobbler/utils.py:248]: Typo:
```suggestion
    Restarts a daemon service independent of the underlining process manager. Currently SupervisorD, systemd and SysV
```

**Review by agraul — APPROVED** [review]
I haven't tried running the code, the logic looks fine to me. Some of the doc strings should probably be updated, I left suggestions in-line.

**Pull request merged**
